Python/ceval.c

```diff
@@ -4066,7 +4066,88 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, PyFrameObject *f, int throwflag)
             CHECK_EVAL_BREAKER();
             DISPATCH();
         }
-
+        case TARGET(CALL_FUNCTION_ADAPTIVE): {
+            SpecializedCacheEntry *cache = GET_CACHE();
+            if (cache->adaptive.counter == 0) {
+                next_instr--;
+                if (_Py_Specialize_CallFunction(BUILTINS(), stack_pointer,
+                    cache->adaptive.original_oparg, next_instr, cache) < 0) {
+                    goto error;
+                }
+                DISPATCH();
+            }
+            else {
+                STAT_INC(CALL_FUNCTION, deferred);
+                cache->adaptive.counter--;
+                oparg = cache->adaptive.original_oparg;
+                JUMP_TO_INSTRUCTION(CALL_FUNCTION);
+            }
+        }
+        case TARGET(CALL_CFUNCTION_FAST): {
+            /* Builtin METH_FASTCALL functions, without keywords */
+            SpecializedCacheEntry *caches = GET_CACHE();
+            _PyAdaptiveEntry *cache0 = &caches[0].adaptive;
+            _PyCallCFunctionCache *cache1 = &caches[-1].call_function;
+            PyObject **callable = &PEEK(cache0->original_oparg + 1);
+            DEOPT_IF(!PyCFunction_CheckExact(*callable), CALL_FUNCTION);
+            PyCFunction cfunc = PyCFunction_GET_FUNCTION(*callable);
+            DEOPT_IF(cfunc != cache1->cfunc, CALL_FUNCTION);
```

Comment:
Rather than caching the exact function, it might make sense to just check whether it is a `METH_FASTCALL` function.
There seems to be no real advantage in specializing on a particular function.